sql/src/test/java/io/crate/planner/TableStatsServiceTest.java

```diff
@@ -0,0 +1,129 @@
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.hamcrest.Matchers;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+
+public class TableStatsServiceTest extends CrateUnitTest  {
+
+    private ThreadPool threadPool;
+
+    @Before
+    public void init() throws Exception {
+        threadPool = new ThreadPool("dummy");
+    }
+
+    @After
+    public void clearThreadPool() throws Exception {
+        threadPool.shutdown();
+        threadPool.awaitTermination(30, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testNumDocs() throws Exception {
+        final AtomicInteger numRequests = new AtomicInteger(0);
+        final TransportSQLAction transportSQLAction = new TransportSQLAction(
+                mock(ClusterService.class),
+                ImmutableSettings.EMPTY,
+                threadPool,
+                mock(Analyzer.class),
+                mock(Planner.class),
+                mock(Provider.class),
+                mock(TransportService.class),
+                mock(StatsTables.class),
+                new ActionFilters(ImmutableSet.<ActionFilter>of()),
+                mock(TransportKillJobsNodeAction.class)
+        ) {
+            @Override
+            protected void doExecute(SQLRequest request, ActionListener<SQLResponse> listener) {
+                Object[] row;
+                if (numRequests.get() == 0) {
+                    row = new Object[] { 2L, "foo", "bar"};
+                } else {
+                    row = new Object[] { 4L, "foo", "bar"};
+                }
+                listener.onResponse(new SQLResponse(
+                        new String[] {"cast(sum(num_docs) as long)", "schema_name", "table_name"},
+                        new Object[][] { row },
+                        new DataType[] {DataTypes.LONG, DataTypes.STRING, DataTypes.STRING},
+                        1L,
+                        1L,
+                        false
+                ));
+                numRequests.incrementAndGet();
+            }
+        };
+
+        TableStatsService statsService = new TableStatsService(
+                ImmutableSettings.EMPTY, threadPool, TimeValue.timeValueMillis(100), new Provider<TransportSQLAction>() {
+            @Override
+            public TransportSQLAction get() {
+                return transportSQLAction;
+            }
+        });
+
+        assertThat(statsService.numDocs(new TableIdent("foo", "bar")), is(2L)); // first call triggers request
+        assertThat(statsService.numDocs(new TableIdent("foo", "bar")), is(2L)); // second call hits cache
+        int slept = 0;
+        while (numRequests.get() < 2 && slept < 1000) {
```

Comment:
should be, the slept < 1000 is just a failsafe in case numRequests never increases (which should after 100 ms are elapsed)